pallets/foreign-investments/src/entities.rs

```diff
@@ -203,89 +118,96 @@ impl<T: Config> InvestmentInfo<T> {
 		investment_id: T::InvestmentId,
 		swapped_pool_amount: T::PoolBalance,
 		swapped_foreign_amount: T::ForeignBalance,
+		pending_foreign_amount: T::ForeignBalance,
 	) -> DispatchResult {
-		self.correlation
-			.increase(swapped_pool_amount, swapped_foreign_amount)?;
+		self.foreign_amount
+			.ensure_add_assign(swapped_foreign_amount)?;
 
-		self.increase_investment(who, investment_id, swapped_pool_amount)
+		if !swapped_pool_amount.is_zero() {
+			T::Investment::update_investment(
+				who,
+				investment_id,
+				T::Investment::investment(who, investment_id)?.ensure_add(swapped_pool_amount)?,
+			)?;
+		}
+
+		if pending_foreign_amount.is_zero() {
+			self.order_id = None;
+		}
+
+		Ok(())
 	}
 
-	/// This method is performed after resolve the swap by cancelling it
-	#[allow(clippy::type_complexity)]
-	pub fn post_increase_swap_by_cancel(
+	/// Decrease an investment taking into account that a previous increment
+	/// could be pending.
+	pub fn cancel(
 		&mut self,
 		who: &T::AccountId,
 		investment_id: T::InvestmentId,
-		swapped_pool_amount: T::PoolBalance,
-		swapped_foreign_amount: T::ForeignBalance,
-	) -> Result<
-		Option<ExecutedForeignDecreaseInvest<T::ForeignBalance, T::CurrencyId>>,
-		DispatchError,
-	> {
-		self.increase_investment(who, investment_id, swapped_pool_amount)?;
+	) -> Result<(T::ForeignBalance, T::PoolBalance), DispatchError> {
+		let swap_id = (investment_id, Action::Investment);
+		let pool_currency = pool_currency_of::<T>(investment_id)?;
 
-		self.decrease_swapped_foreign_amount
-			.ensure_add_assign(swapped_foreign_amount)?;
+		let cancel_pool_amount = T::Investment::investment(who, investment_id)?;
+		T::Investment::update_investment(who, investment_id, Zero::zero())?;
 
-		let no_pending_decrease = self.pending_decrease_swap(who, investment_id)?.is_zero();
-		if no_pending_decrease && !self.decrease_swapped_foreign_amount.is_zero() {
-			return Ok(Some(ExecutedForeignDecreaseInvest {
-				foreign_currency: self.foreign_currency,
-				amount_decreased: sp_std::mem::take(&mut self.decrease_swapped_foreign_amount),
-				amount_remaining: self.remaining_foreign_amount(who, investment_id)?,
-			}));
-		}
+		if self.foreign_currency != pool_currency {
+			let increase_foreign = match self.order_id {
+				Some(order_id) => {
+					let increase_foreign = cancel_swap::<T>(who, swap_id, &order_id)?.into();
```

Comment:
> How do we know at this point that if there exists an order

We now there is an order because we have stored the `order_id`

> it is an order from foreign to pool which we can cancel in cancel_swap?

Yes

> It could also be an order which swaps pool to foreign from a previous cancel?

No, because you can not call cancel again if there is a pending cancel in progress